substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/DebugHelper.java

```diff
@@ -281,6 +295,493 @@ public static void printLocationInfo(@SuppressWarnings("unused") IsolateThread t
             SubstrateDiagnostics.printLocationInfo(Log.log(), mem, true, true);
         }
     }
+    public static class IdeDebugHelper {
+        /**
+         * Convert instance of java.lang.String to utf8 c-string.
+         * @param thread
+         * @param ptr pointer to instance of java.lang.String
+         * @return c-string
+         */
+        @CEntryPoint(name = "svm_dbg_string_to_utf8", include = IncludeDebugHelperMethods.class, publishAs = Publish.SymbolOnly)
+        @CEntryPointOptions(prologue = SetThreadAndHeapBasePrologue.class)
+        public static CCharPointer stringToUtf8(@SuppressWarnings("unused") IsolateThread thread, Pointer ptr) {
```

Comment:
Let me be more specific how this (and many of your other methods) may break.

Scenario A
- Thread A tries to initiate a safepoint.
- Thread B hits a gdb breakpoint. All threads are paused in gdb.
- Thread B calls any of your methods that allocates Java memory.
- Thread B needs to initiate a garbage collection (GC) and deadlocks because thread A is already trying to initiate a safepoint but is blocked in gdb.

Scenario B
- Thread A executes some VM internal code.
- Thread B hits a gdb breakpoint. All threads are paused in gdb.
- Thread B allocates Java memory and needs to refill its thread-local allocation buffer (TLAB). The TLAB refill deadlocks because it tries to acquire a lock that is already locked by thread A in some VM internal code.

Scenario C
- The user sets a breakpoint in some JDK code that is used VM-internally as well.
- Thread A hits that breakpoint and gets paused in gdb while executing VM-internal code that must be allocation free.
- Thread A calls one of your methods that allocates Java heap memory and either crashes right away or destroys VM-internal state.

Scenario D
- Thread A hits a breakpoint and gets paused in gdb.
- Thread A calls one of your methods that allocates Java heap memory and the allocation throws an `OutOfMemoryError`.
- The process crashes because the `@CEntryPoint` uses an exception handler that kills the process by default.

Scenario E (with dedicated VM operation thread)
- Thread A hits a gdb breakpoint. All threads are paused in gdb.
- Thread A calls one of your methods that allocates Java memory.
- Thread A needs to initiate a GC and deadlocks because the dedicated VM operation thread is paused as well, so no one is going to execute the requested GC.

Scenario F
- Thread A hits a gdb breakpoint. Only thread A is paused in gdb all other threads are still running.
- Thread A calls `svm_dbg_allocate_object(...)`, which returns a raw pointer to a newly allocate object (`obj1`).
- Thread A calls `svm_dbg_allocate_object(...)` a second time. This time, the allocation needs a GC, which moves `obj1` that was returned by the first call. If this outdated pointer is access, it may result in crashes or may return invalid data.

Those are the scenarios that come to my mind without thinking about it too much in detail. I am definitely sure that there are more complex scenarios as well that I completely missed in this rather quick analysis and that will cause similar or worse issues.

Please provide the additional information that I requested yesterday so that I can try to come up with a design that doesn't have those issues.